File: create-a-deep-flattening-tool.py
def steamroll_array(arr):
    flattened = []

    def flatten(item):
        if isinstance(item, list):
            for sub_item in item:
                flatten(sub_item)  # 再帰呼び出し
        else:
            flattened.append(item)

    for element in arr:
        flatten(element)

    return flattened

#Alt

def steamroll_array_gen(arr):
    def flatten_gen(item):
        if isinstance(item, list):
            for sub_item in item:
                yield from flatten_gen(sub_item)
        else:
            yield item

    return list(flatten_gen(arr))

#yield from

def steamroll_gen(arr):
    for item in arr:
        if isinstance(item, list):
            # サブジェネレータに処理を丸投げ（委譲）
            yield from steamroll_gen(item)
        else:
            yield item

nested_list = [1,[2,[3,4]],5]
gen = steamroll_gen(nested_list)
for val in gen:
    print(val, end=" ")

###if isinstance(item,list):
    ###for sub_item in streamroll_gen(item):
       ### yield sub_item

###if isinstance(item,list):
   ### yield from streamroll_gen(item)

#

def streamroll_array_stack(arr):
    stack=list(arr)
    result =[]

    while stack:
        item = stack.pop()

        if isinstance(item,list):
            stack.extend(item)
        else:
            result.append(item)
    result.reverse()
    return result

print(streamroll_array_stack([1,{},[3,[[4]]]]))

#

def read_hug_file(file_path):
    with open(file_path,"r")as f:
        for line in f:
            yield line

#for line in read_hug_file("huge_file.txt"):
#    process(line)

    
import tracemalloc

tracemalloc.start()

snapshot = tracemalloc.take_snapshot()
top_stats = snapshot.statistics("lineno")
for stat in top_stats[:5]:
    print(stat)

File: create-a-deep-flattening-tool.js
function steamrollArray(arr) {
  const flattened = [];

  function flatten(item) {
    if (Array.isArray(item)) {
      // 
      for (let i = 0; i < item.length; i++) {
        flatten(item[i]);
      }
    } else {
      flattened.push(item);
    }
  }

  for (let i = 0; i < arr.length; i++) {
    flatten(arr[i]);
  }

  return flattened;
}

//alt

function stramrollArray(arr) {
    return arr.reduce((acc, val) => {
        return acc.concat(Array.isArray(val)? stramrollArray(val): val);
    }, []);
}

//

function streamrollArrayStack(arr){
  const stack = [...arr];
  const result = [];
  while(stack.length > 0){
    const next = stack.pop();

    if(Array.isArray(next)){
      stack.push(...next);
    } else{
      result.push(next);
    }
  }
  return result;
}

console.log(streamrollArrayStack([1,{},[3,[[4]]]]));

//


import readline from 'readline';
import fs from 'fs';

const rl = readline.createInterface({ input: fs.createReadStream('hug_file.txt')});
rl.on('line',(line)=> {
  //O(1);
});

/* node --max-old-space-size-8192 app.js */

//

import fs from 'node.fs';
import { pipeline } from 'node:stream/promises';
import { Transform } from 'node:stream';
import realine from 'node:readline';

const filterStream = new Transform({
  transform(chunk, encoding, callback) {
    const line = chunk.toString();

    if (line.includes('ERROR')){
      this.push(line + '\n');
    }
    callback();
  }
});

async function processHugeFile(){
  try {
    await pipeline(
      fs.createReadStream('input_huge_log.txt'),
      filterStream,
      fs.createWriteStream('output_errors.txt')
    );

  console.log('処理が正常に完了しました。');  
  } catch {
  console.error('ストリーム処理中にエラーが発生しました:', err);
  }
}

//

import fs from 'node:fs';
import { pipeline } from 'node:strea/promises';

async function* transformData(source){
  for await (const chunk of source) {
    const text= chunk.toString().toUpperCase();
    yield text;
  }
}

async function run() {
  try {
    await pipeline(
      fs.createReadStream('large_file.txt'),
      transformData,
      fs.createWriteStream('transformed_file.txt')
    );
    cconsole.log('パイプライン処理成功！');
  } catch (err) {
    console.error('処理失敗:', err);
  }
}

run();